stdlib/src/builtin/string.mojo

```diff
@@ -2109,3 +2205,83 @@ fn _calc_format_buffer_size[type: DType]() -> Int:
         return 64 + 1
     else:
         return 128 + 1  # Add 1 for the terminator
+
+
+# ===----------------------------------------------------------------------===#
+# Format method structures
+# ===----------------------------------------------------------------------===#
+
+
+@value
+struct _FormatCurlyEntry:
+    """
+    Internally used by the `format()` method.
+
+    Purpose of the variant types:
+
+    - `String` for **kwargs
+
+        ```mojo
+        "{message}".format(message="hello")
+        ```
+
+        value field contains `message`
+
+    - `Int` for manual indexing of *args
+
+        ```mojo
+        "{0}".format("hello")
+        ```
+
+        value field contains `0`
+
+    - `NoneType` for automatic indexing of *args
+
+        ```mojo
+        "{}".format("hello")
+        ```
+
+        value field contains `None`
+
+    `*_curly` are the indexes of the enclosing curlies
+
+    This struct only supports theses:
+
+    `{number|name|nothing}`
+
+    """
+
+    var first_curly: Int
+    var last_curly: Int
+    var value: Variant[String, Int, NoneType]
```

Comment:
Can you please go through and update the code to ensure that complexity from the keyword handling is removed?

Reply:
Yes it should, but having the one for keyword let's us raise an error if the user do:
```mojo
"{} {foo}".format("bar")
#❌ Index foo not in kwargs
```
Even if we do not support keyword args, the error need to be raised no ?
Maybe we should parametrize `_FormatCurlyEntry.create_entries[raises_on_kwargs:Bool]` ?

Reply:
Why not just raise directly from `create_entries`? This `create_entries` function is what performs the actual parsing, so the rules of what the fields/entries can look like should be defined and checked here, for example `raise "Cannot both use manual and automatic indexing"`. Right now, the fields cannot be strings yet, which you can enforce here. Basically, I want to make we check as much as we can here before handing over the entry list to format later.

We don't have the ability to raise parametrically yet, but that's okay, just mark it raising.

Reply:
Yes, we are now raising from `create_entries` :+1: